aioxmpp/muc/service.py

```diff
@@ -618,18 +633,28 @@ def _match_tracker(self, message):
         try:
             tracker = self._tracking_by_id[message.id_]
         except KeyError:
+            key = message.from_, _extract_one_pair(message.body)
```

Comment:
Right, sorting is also O(n) (per message).

We should probably simply drop the ``from`` address from the dict and always compare with ``self._this_occupant.conversation_jid``. This helps to defeat races by nickname, saves space *and* gives us an O(1) check to weed out the majority of messages.

Then it doesn’t harm *that* much if we actually do the linear scan, either in the ``_extract_one_pair`` or when searching for a match.

Does that sound right?

(Of course, a malicious MUC could still reflect messages with a whole lot of ``<body/>`` elements, but that will only affect this code for those which are coming from our address, making the attack obvious to the user and the MUC useless.)